### PR TITLE
Remove futures_[un]ordered functions

### LINKED ISSUE #1477: Iterators for FuturesUnordered are private
The `IterMut` and `IterPinMut` types returned by [`futures::stream::FuturesUnordered::iter_mut`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.13/futures/stream/struct.FuturesUnordered.html#method.iter_mut) and [`futures::stream::FuturesUnordered::iter_pin_mut`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.13/futures/stream/struct.FuturesUnordered.html#method.iter_pin_mut) are not exposed.

### PR DESCRIPTION
**Pull request by taiki-e**
Closes #1477

### CODE REVIEW

File: futures-util/src/stream/mod.rs [hunk 1]
```diff
@@ -128,7 +128,7 @@ cfg_target_has_atomic! {
     #[cfg(feature = "alloc")]
     mod futures_unordered;
     #[cfg(feature = "alloc")]
-    pub use self::futures_unordered::{futures_unordered, FuturesUnordered};
+    pub use self::futures_unordered::{futures_unordered, FuturesUnordered, IterMut, IterPinMut};
```
Comment (cramertj) [inline futures-util/src/stream/mod.rs]: If we're going to expose these at top-level, IMO they need less generic names (`FuturesUnorderedIterMut` or something like that)

Comment (taiki-e) [reply futures-util/src/stream/mod.rs]: Indeed!

File: futures-util/src/stream/mod.rs [hunk 2]
```diff
@@ -126,8 +126,9 @@ cfg_target_has_atomic! {
     pub use self::futures_ordered::{futures_ordered, FuturesOrdered};
 
     #[cfg(feature = "alloc")]
-    mod futures_unordered;
+    pub mod futures_unordered;
```
Comment (taiki-e) [inline futures-util/src/stream/mod.rs]: It was not possible to overwrite `FuturesUnordered::iter_pin_mut` and `FuturesUnordered::iter_mut` return types in the document simply by renaming on export, so I changed the module itself to be exposed like [`proc_macro2::token_stream`](https://docs.rs/proc-macro2/0.4/proc_macro2/token_stream/index.html).

Comment (taiki-e) [reply futures-util/src/stream/mod.rs]: futures 0.1 also exposed `futures_unordered` module: https://docs.rs/futures/0.1.25/futures/stream/futures_unordered/index.html

Comment (taiki-e) [reply futures-util/src/stream/mod.rs]: I overlooked the discussion in https://github.com/rust-lang-nursery/futures-rs/issues/793.

**Comment by Nemo157** [comment]
Personally I'd go for [@cramertj's proposal](https://github.com/rust-lang-nursery/futures-rs/issues/793#issuecomment-368576067) in the linked issue, remove the `futures_unordered` function and use `FuturesUnordered::from_iter` instead. Then the iterators can be namespaced away since they're going to be very rarely used types.

Probably good to wait and see if @cramertj's opinion has changed since then though.

**Comment by cramertj** [comment]
I don't have a strong opinion either way. @Nemo157 feel free to review this and merge if it matches your expectations.

**Comment by Nemo157** [comment]
Thanks, this looks great.

**Comment by Benjamin-L** [comment]
I don't know if I'm missing something, but what should people be doing instead of `futures_ordered` now?

**Comment by Nemo157** [comment]
@Benjamin-L you can use the [`FromIterator` implementation of `FuturesOrdered`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.13/futures/stream/struct.FuturesOrdered.html#impl-FromIterator%3CFut%3E). Either by using `some_futures.into_iter().collect::<FuturesOrdered>()` or `use std::iter::FromIterator; FuturesOrdered::from_iter(some_futures);` (I really wish that `FromIterator` made it into the prelude so that that could be the goto recommendation).

**Comment by Arnavion** [comment]
Needs to be mentioned in the alpha.14 changelog since it's a breaking change. (Took me a bunch of git-blame-ing to discover this.)

**Pull request merged**